internal/olm/operator/bundle/install.go

```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+	if err != nil {
+		return "", fmt.Errorf("error converting the declarative config to string: %v", err)
+	}
+
+	if content == "" {
+		return "", errors.New("file based catalog contents cannot be empty")
+	}
+
+	log.Infof("Generated a valid File-Based Catalog")
+
+	return content, nil
+}
+
+// addBundleToIndexImage adds the bundle to an existing index image if the bundle is not already present in the index image.
+func addBundleToIndexImage(ctx context.Context, indexImage string, bundleDeclConfig *declarativeconfig.DeclarativeConfig) (*declarativeconfig.DeclarativeConfig, error) {
+	log.Infof("Rendering a File-Based Catalog of the Index Image %q", indexImage)
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: []string{indexImage},
+	}
+
+	imageDeclConfig, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+	if err != nil {
+		return nil, fmt.Errorf("error rendering the index image %q: %v", indexImage, err)
+	}
+
+	var isPackagePresent, isBundlePresent, isChannelPresent bool
+	for _, pkg := range imageDeclConfig.Packages {
+		if pkg.Name == bundleDeclConfig.Packages[0].Name {
+			isPackagePresent = true
+			break
+		}
+	}
+
+	for _, bundle := range imageDeclConfig.Bundles {
+		if bundle.Name == bundleDeclConfig.Bundles[0].Name && bundle.Package == bundleDeclConfig.Bundles[0].Package {
+			isBundlePresent = true
+			break
+		}
+	}
+
+	for _, channel := range imageDeclConfig.Channels {
+		if channel.Name == bundleDeclConfig.Channels[0].Name && channel.Package == bundleDeclConfig.Bundles[0].Package {
+			isChannelPresent = true
+			break
+		}
+	}
+
+	// insert bundle to index if package/bundle/channel is not present on the index;
+	// and insert only bundle declarative config channel to index when both bundle/package are present but not the channel.
+	if !isPackagePresent && !isBundlePresent && !isChannelPresent {
+		imageDeclConfig.Bundles = append(imageDeclConfig.Bundles, bundleDeclConfig.Bundles[0])
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
+		imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others...)
+	} else if isPackagePresent && isBundlePresent && !isChannelPresent {
+		// if the bundle already exists in a different channel,
+		// we already have the bundle and package blobs in the FBC, in which case,
+		// we need to reference it in the channel we want it to be in.
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+	} else if isBundlePresent {
+		return nil, fmt.Errorf("bundle %q already present in the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	} else {
+		return nil, fmt.Errorf("cannot add bundle %q to the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	}
+
+	log.Infof("Inserted the new bundle %q into the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+
```

Comment:
I _think_ we can significantly simplify this insertion logic.

First, we can iterate `imageDeclConfig.Bundles`. If we find a bundle that matches the package name and bundle name of `bundleDeclConfig`, we stop and error out with "bundle already exists in the catalog image".

If we get through that loop without erroring, we can be sure that the bundle isn't in the catalog, so then we know we're adding at least the bundle. Then we have the separate question of whether we need to add/update the package and channel blob associated with that bundle.

It seems like the existing logic actually permits `run bundle` to succeed even if the catalog image already contains the bundle (i'm thinking lines 258-262), which I think we said should not be allowed.

Reply:
The current logic will succeed if the bundle is present but the channel is _not_ present. This is a use case we want to support, right? For any other case where the bundle exists, the current logic will error out, as we want it to. If we don't want to support the use case above, then I agree with your logic. 

You mentioned **updating** the package/channel. In what scenario would be be _updating_ anything in the FBC? It seems like the only thing we need to do throughout this run bundle call is add items to generate an FBC, and not really modify any package/channel inside. Or have I misunderstood the use of that word in this context?

Reply:
@joelanford could you please clarify whether we need to support the use-case of updating the channel/package if the bundle is already present. 

The reason why we are permitting `run bundle` to succeed is that a bundle can exist in the index image but in a different channel and not in the channel we want to add it in - in this case, the bundle would already exist in the index but not the channel, so we need to append the channel from `bundleDeclConfig` to `imageDeclConfig`. Shouldn't we be doing this? 

I made this change based on your earlier review comment [here](https://github.com/operator-framework/operator-sdk/pull/5671#discussion_r853416273) although I may have misunderstood it, so looking to get these clarifications from you on whether we could keep this existing logic or change it. Please let me know.

Reply:
@joelanford Any thoughts on this? Can we proceed with the assumption that we need to update the channel if the bundle is already present in the index?

Reply:
In sqlite, if you try to `opm add` a bundle that's already in the DB, regardless of which channels its in, `opm add` will fail. So `operator-sdk run bundle` inherits that behavior when dealing with sqlite catalogs. 

I think we should have equivalent logic for FBC -- basically if the bundle exists in the catalog, we'll fail no matter the channel differences.

Reply:
@joelanford I agree. That makes sense. I've simplified the insertion logic to fail if bundle is present on index and add all three (channel/package/bundle) if none of them exist on the index.

Reply:
If the bundle's package is already present, but the bundle is not, it seems like this logic causes a failure in that case?

It seems like once we've determined that the bundle is not present in the catalog, we should definitely be adding the bundle.

This [comment](https://github.com/operator-framework/enhancements/pull/112#discussion_r866025776) sums up what it seems like this logic needs to be:

> 1.  Does this bundle name/package already exist in the catalog? If so, fail.
> 2.  Does this bundle's package already exist in the catalog? If so, skip package blob generation
> 3.  Generate a channel blob using a name that does not conflict with an existing channel in the bundle's package. Maybe hardcoded as operator-sdk-run-bundle?
> 4.  Generate the bundle blob.
> 5. Insert the generated FBC into <fbcRootDir>/operator-sdk-run-bundle/extra.yaml and run the catalog image.

Since we aren't yet doing the "only add the extra FBC" stuff yet (but we plan to), I think the logic should fail out if the generated channel name conflicts with an existing channel name. And I think we should pick a channel name that is exceedingly unlikely to conflict (e.g. `operator-sdk-run` or something along those lines)

So, summing up:
1. If the rendered bundle is already present -- fail
2. If the generated channel is already present -- fail
3. If the generated package is already present - skip package generation
4. Add the bundle and channel -- and optionally the package (based on step 3).

Reply:
@joelanford, the latest commit reflects this logic, PTAL.